### DISCUSSION TITLE
Is this lib missing a dependency on CombineSchedulers?

### DISCUSSION THREAD

**Discussion by ryanmeisters** [pointfreeco/swift-sharing#34] [Q&A]
### Description

Howdy! We have a large project with tons of modules. When updating to TCA 1.17.0 I'm getting these linking errors when xcode tries to link Sharing:

```
Showing All Messages
Undefined symbol: CombineSchedulers.AnyScheduler.now.getter : A

Undefined symbol: static (extension in CombineSchedulers):CombineSchedulers.AnyScheduler< where A == (extension in Dispatch):__C.OS_dispatch_queue.SchedulerTimeType, B == (extension in Dispatch):__C.OS_dispatch_queue.SchedulerOptions>.immediate.getter : CombineSchedulers.AnyScheduler<(extension in Dispatch):__C.OS_dispatch_queue.SchedulerTimeType, (extension in Dispatch):__C.OS_dispatch_queue.SchedulerOptions>

Undefined symbol: type metadata accessor for CombineSchedulers.AnyScheduler

Undefined symbol: nominal type descriptor for CombineSchedulers.AnyScheduler

Undefined symbol: protocol conformance descriptor for CombineSchedulers.AnyScheduler<A, B> : Combine.Scheduler in CombineSchedulers

Linker command failed with exit code 1 (use -v to see invocation)
```

I'm wondering if Sharing is missing a dependency on CombineSchedulers, e.g. [this code here](https://github.com/pointfreeco/swift-sharing/blob/464a3292d4c9630d33b8a505e4bb058caf57b8fc/Sources/Sharing/SharedKeys/FileStorageKey.swift#L343) looks like `.immediate` comes from there? 

Our project is complex so this could be a me problem, but wanted to make sure i'm not crazy.

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-sharing/issues) or [discussion](https://github.com/pointfreeco/swift-sharing/discussions).

**Comment by stephencelis** [comment]
@ryanmeisters swift-sharing depends on swift-dependencies, which depends on combine-schedulers, so I don't think this is the problem you are seeing. I'm going to convert to a discussion till we can determine this is a bug in the library. Can you share a repro so we can see if this is something to do with your project vs. our library?

**Reply by ryanmeisters** [reply]
Thanks for the fast reply! I will try to narrow it down to something shareable today 🙏🏼

**Reply by ryanmeisters** [reply]
Ok so i'm actually able to repo really easily.

1. Fresh ios app project (xcode 16.1)
2. Add TCA 1.17
3. Create a framework target
4. Add ComposableArchitecture to "Frameworks and Libraries" for both app target and framework target
5. Build and see linking errors

Here's a zip!
 
[TCA17Test2.zip](https://github.com/user-attachments/files/18103707/TCA17Test2.zip)

**Reply by ryanmeisters** [reply]
Hmm I also just tried the same but with TCA 1.16.1 and got this error instead:

In SwiftNavigation's UIBinding.swift
```
External macro implementation type 'PerceptionMacros.PerceptibleMacro' could not be found for macro 'Perceptible()'; plugin for module 'PerceptionMacros' not found
```

**Reply by stephencelis** [reply]
@ryanmeisters What happens if you just link to your framework target? That should make TCA available to your app target anyway, so no need to double-link.

**Reply by ryanmeisters** [reply]
Yeah that works for the example project. I'm not sure it's an option for our main project though (200+ static framework modules), but I can try. 

Fwiw, the configuration above does compile without issue with TCA 1.11.2 (the version that we're currently stuck on). Don't want to take too much of your time. If we're the only ones with this problem then it probably means we're doing something weird.

**Reply by bedrock-drivet-externe** [reply]
Hi folks,

I'm Damien, working at Bedrock streaming (Lyon, France) and we're facing the exact same issue with our App.

Some context regarding our situation: when doing some R&D to check if our codebase was ready for the migration from **Xcode 16.0** to **Xcode 16.1** (and later **Xcode 16.2**), we found out that we had a singular compilation issue with `swiftui-navigation` which was fixed in the `swift-navigation` new repo, so I tried to update to the latest version of `swift-composable-architecture` (`1.3.0` -> `1.17.0`).

I first had another issue with the `@Perceptible` macro missing when compiling but with some cleanup I made it disappear (I'll try to reproduce it with the test project supplied above).

Like @ryanmeisters, we have a pretty similar architecture with 75+ frameworks, a few of them are importing `swift-composable-architecture` and because of some specific needs, we also have an internal **Cocoapods** plugin that adds SPM dependencies to our frameworks & apps.

But the solution mentioned above does not work in our case (it does work for the test project), removing TCA-related packages from the main App does not solve the problem for us.

I'll be working on the issue till I can find a suitable solution, if you want me to run some tests, tweak our configuration, etc. feel free to ping me here.

**Reply by bedrock-drivet-externe** [reply]
For the record, the release notes from both **16.1** and **16.2** contains nothing related to compilation changes.

https://developer.apple.com/documentation/xcode-release-notes/xcode-16_1-release-notes

https://developer.apple.com/documentation/xcode-release-notes/xcode-16_2-release-notes

**Reply by mbrandonw** [reply]
Hi @ryanmeisters and @bedrock-drivet-externe, I suppose there is a chance that Xcode is getting more strict with transitive dependencies. Combine Schedulers should be coming from Dependencies just fine, but just in case I have created a branch, [`explicit-combine-schedulers`](https://github.com/pointfreeco/swift-sharing/compare/explicit-combine-schedulers?expand=1), that adds the dependency explicitly. Would you be able to point to that and see if it fixes things?

**Reply by bedrock-drivet-externe** [reply]
Hi @mbrandonw,

Thanks for the quick answer 👍 Being explicit with the dependency did fix something as the `CombineSchedulers` error is now gone. But the `@Perceptible` macro compilation error is back in the game 😞

|![Screenshot 2024-12-18 at 09 54 21](https://github.com/user-attachments/assets/85504406-82b4-42a8-89e3-dbb35838321e)|
|---|
|App target with explicit sharing dependency|

|![Screenshot 2024-12-18 at 09 54 09](https://github.com/user-attachments/assets/113644b9-dca8-4b22-a2d4-c5ee65eac2dd)|
|---|
|App target with error|

|![Screenshot 2024-12-18 at 09 54 12](https://github.com/user-attachments/assets/f23f6d68-2536-4b0a-97eb-1a85dd505c9a)|
|---|
|Framework with explicit sharing dependency added for testing|

**Reply by bedrock-drivet-externe** [reply]
For reference, here's the compilation error site.

![Screenshot 2024-12-18 at 10 03 08](https://github.com/user-attachments/assets/b328676a-53c5-4ebd-8e4b-c2ab12167776)

**Reply by bedrock-drivet-externe** [reply]
After a lot of searching on the interwebs, I tried a few solution but unfortunately, none of them worked so far.

https://forums.swift.org/t/how-to-import-macros-using-methods-other-than-swiftpm/66645/40
https://github.com/pointfreeco/swift-composable-architecture/discussions/3419
https://github.com/pointfreeco/swift-composable-architecture/discussions/2639
https://stackoverflow.com/questions/77386744/swift-macros-external-macro-implementation-type-could-not-be-found

**Reply by mbrandonw** [reply]
Hi @bedrock-drivet-externe, are you able to provide a project that demonstrates the problem you are having?

**Reply by mbrandonw** [reply]
@bedrock-drivet-externe: I believe this will also fix the macro errors you are seeing: https://github.com/pointfreeco/swift-navigation/pull/256

**Reply by bedrock-drivet-externe** [reply]
@mbrandonw https://github.com/pointfreeco/swift-navigation/pull/256 did fix the build issue with the `@Perceptible` macro. Thanks a lot for the investigation and the fixes on both `swift-sharing` and `swift-navigation`.

Interestingly, I did some testing with **Tuist** to see how it would integrate the same project architecture and the only time it "worked" was by making the `FeatureA` framework a static framework with `MACH_O_TYPE` set to `Static library` so it matches your statement about **Xcode 16.1+** changing something about dynamic framework.

When can we expect those fixes to be fully released? I imagine that there will be some internal testing before they go live.
